sys/kern/kmem.c

```diff
@@ -74,21 +74,20 @@ vm_page_t *kva_find_page(vaddr_t ptr) {
   return NULL;
 }
 
-void kva_unmap(vaddr_t ptr, size_t size) {
-  assert(page_aligned_p(ptr) && page_aligned_p(size));
+void kva_unmap(vaddr_t va, size_t size) {
+  assert(page_aligned_p(va) && page_aligned_p(size));
 
-  kasan_mark_invalid((void *)ptr, size, KASAN_CODE_KMEM_FREED);
+  kasan_mark_invalid((void *)va, size, KASAN_CODE_KMEM_FREED);
 
-  vaddr_t va = (vaddr_t)ptr;
-  vaddr_t end = va + size;
-  while (va < end) {
+  size_t end = va + size;
+  size_t pgsz;
+  for (; va < end; va += pgsz) {
     vm_page_t *pg = kva_find_page(va);
     assert(pg != NULL);
-    va += pg->size * PAGESIZE;
+    pgsz = pg->size * PAGESIZE;
+    pmap_kremove(va, pgsz);
```

Comment:
It would be more efficient to unmap all the pages at once before the loop. Can we do that here?

Reply:
Ok. I reverted to the original structure of this function.

However... there's something fishy here. I cannot perform `pmap_kremove` before the loop, since I use `va -> pa` mapping to extract `vm_page_t` addresses. So there's a window of time when formally pages are free, but still visible through kernel virtual address space.

Ideally I would have to:
 * collect all pages to free on a list,
 * unmap the region with `pmap_kremove`,
 * and free the pages with `vm_pagelist_free`. 

The only problem is that the last function does not exists. Should I add it ?

Reply:
Seems like the answer is yes.